File: Ser2023/Company/Company.Web/Modules/ServerTypes/Store/ProductsForm.ts
import { LookupEditor, StringEditor, ImageUploadEditor, BooleanEditor, DecimalEditor, IntegerEditor, PrefixedContext } from "@serenity-is/corelib";
import { MeasuresDialog } from "@/Store/Measures/MeasuresDialog";
import { ProductDetailsEditor } from "@/Store/ProductDetails/ProductDetailsEditor";
import { initFormType } from "@serenity-is/corelib/q";

export interface ProductsForm {
    CategoryId: LookupEditor;
    ProductName: StringEditor;
    ProductCode: StringEditor;
    ProductBarcode: StringEditor;
    ProductLabel: StringEditor;
    MeasureId: LookupEditor;
    ProductImage: ImageUploadEditor;
    Discontinued: BooleanEditor;
    UnitPrice: DecimalEditor;
    QuantityPerUnit: StringEditor;
    DetailList: ProductDetailsEditor;
    UnitsInStock: IntegerEditor;
    UnitsOnOrder: IntegerEditor;
    ReorderLevel: IntegerEditor;
}

export class ProductsForm extends PrefixedContext {
    static formKey = 'Store.Products';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!ProductsForm.init)  {
            ProductsForm.init = true;

            var w0 = LookupEditor;
            var w1 = StringEditor;
            var w2 = ImageUploadEditor;
            var w3 = BooleanEditor;
            var w4 = DecimalEditor;
            var w5 = ProductDetailsEditor;
            var w6 = IntegerEditor;

            initFormType(ProductsForm, [
                'CategoryId', w0,
                'ProductName', w1,
                'ProductCode', w1,
                'ProductBarcode', w1,
                'ProductLabel', w1,
                'MeasureId', w0,
                'ProductImage', w2,
                'Discontinued', w3,
                'UnitPrice', w4,
                'QuantityPerUnit', w1,
                'DetailList', w5,
                'UnitsInStock', w6,
                'UnitsOnOrder', w6,
                'ReorderLevel', w6
            ]);
        }
    }
}

[MeasuresDialog]; // referenced types

File: Ser2023/Store/Store.Web/Modules/Store/Products/ProductsForm.cs

namespace Store.Store.Forms
{
    using Serenity.ComponentModel;
    using Serenity.Web;
    using System;
    using System.Collections.Generic;
    using System.ComponentModel;

    [FormScript("Store.Products")]
    [BasedOnRow(typeof(Entities.ProductsRow), CheckNames = true)]
    public class ProductsForm
    {
        [Category("General")]
        [HalfWidth]
        public Int32 CategoryId { get; set; }
        [HalfWidth]
        public String ProductName { get; set; }
        [HalfWidth]
        public String ProductCode { get; set; }
        [HalfWidth]
        public String ProductBarcode { get; set; }
        [HalfWidth]
        public String ProductLabel { get; set; }
        [HalfWidth]
        public Int32 MeasureId { get; set; }
        public String ProductImage { get; set; }
        public Boolean Discontinued { get; set; }

        [Category("Pricing")]

        [HalfWidth]
        public Decimal UnitPrice { get; set; }
        [HalfWidth]
        public String QuantityPerUnit { get; set; }


        [Category("Product Details")]
        [ProductDetailsEditor]
        public List<Entities.ProductDetailsRow> DetailList { get; set; }

        [Category("Status")]
        [HalfWidth]
        public Int16 UnitsInStock { get; set; }
        [HalfWidth]
        public Int16 UnitsOnOrder { get; set; }
        [HalfWidth]
        public Int16 ReorderLevel { get; set; }
    }
}